### PR TITLE
Common: add YoloV3 network

### PR DESCRIPTION
**Pull request by jochem-brouwer**
This PR is aimed towards the eip2718-eip2930 branch and creates the YoloV3 block in Common. We successfully import the block, but sadly it does not seem we are connecting to the network. My hunch is, that this is because we don't have the right fork hashes.

[Genesis block source](https://gist.github.com/holiman/c6ed9269dce28304ad176314caa75e97)
[Genesis block as reported by web3](https://gist.github.com/ryanschneider/18d1bb10354c5d3e4d3de772cda6e70e)

The goal of this PR is to successfully connect and sync with YoloV3 (which is also why it is targeted towards the eip2718-eip2930 branch, because we need support for these EIPs in order to sync).

### CODE REVIEW

File: packages/common/src/chains/index.ts [hunk 1]
```diff
@@ -15,13 +16,15 @@ export function _getInitializedChains(customChains?: Chain[]) {
     '4': 'rinkeby',
     '42': 'kovan',
     '5': 'goerli',
+    '34180983699157880': 'yolov3',
```
Comment (jochem-brouwer) [inline packages/common/src/chains/index.ts:19]: I am assuming that these numbers are supposed to be the chain id?

File: packages/common/src/genesisStates/index.ts [hunk 2]
```diff
@@ -7,12 +7,14 @@ const genesisStates: genesisStatesType = {
     '4': 'rinkeby',
     '42': 'kovan',
     '5': 'goerli',
+    '34180983699157880': 'yolov3',
```
Comment (jochem-brouwer) [inline packages/common/src/genesisStates/index.ts:10]: Same here, these should be chain ids right?

File: packages/common/src/chains/index.ts (hunk 1 above)
Comment (holgerd77) [reply packages/common/src/chains/index.ts:19]: Yes.

File: packages/common/src/genesisStates/index.ts (hunk 2 above)
Comment (holgerd77) [reply packages/common/src/genesisStates/index.ts:10]: Yes.

**Comment by holgerd77** [comment]
Will give this a short look now. We also still have this problem in the client (at least I am 80% sure on this), that the first 1-2 bootnode connections are (often? always?) not successful due to all the services not yet set up properly, this might also be the cause for the problems. Extremely hard to debug though (at least I wasn't successful on this yet).

We'll see. 😄

**Comment by holgerd77** [comment]
Phew, had to fix 3-4 (!) partly serious bugs for this in a row (but luckily I had a streak here in this case 😁).

For the `forkHash` there is an internal helper function in `Common` - `calcForkHash()` - which can be used to calculate the correct forkhashes. I spent some longer time to implement the algorithm correctly there, so I think this can normally be trusted. This method actually also jumps in when there is no forkhash provided in the chain files, then the forkhash is just always re-calculated on the fly, adding this to the files is mainly for performance reasons (we might as well just expose `_calcForkHash()` publicly, likely useful for others in other contexts as well).

The second "bug" is a bit eyebrow raising, I noticed that on a received ETH status reply (for completely unknown reason I once managed to get one before the bug fix 😜 ), the next forkblock by geth - if unknown - is apparently announced as the empty buffer and not the zero-Buffer, e.g. this `devp2p:eth Received STATUS message from 3.9.20.133:30303: : [V:64, NID:34180983699157880, TD:157509, BestH:bdf8438, GenH:f1f2876, ForkHash: 0xa308affe, ForkNext: NaN] +1ms` is a respective received status message. This is actually deviating from the spec (or can this also be some RLP encoding thing with zero vs empty?) - since the spec is saying "FORK_NEXT: Block number (uint64) of the next upcoming fork, or 0 if no next fork is known." [here](https://eips.ethereum.org/EIPS/eip-2124#specification). Anyhow, I changed this to match and this worked. Might open a PR towards the EIP with some suggestion for clarification.

Last thing: also stumbled upon the high chainID - this time in the client - could be fixed relatively easy.

Anyhow: **this is now working** (so: the connection part 😋 ) 🎉 , so: happy debugging! 👍 

Cool cool work @jochem-brouwer, so great that we are finally here now! 😄

**Comment by jochem-brouwer** [comment]
Is there a problem in client with the high chain id? The number is high, but not so high that we run into rounding errors.

If we are connecting now does that also mean we are syncing?

**Comment by jochem-brouwer** [comment]
Looks like we are actually syncing some blocks?

Then I get an error

```
ERROR [02-27|13:23:27] Error: Number can only safely store up to 53 bits
    at assert (/Volumes/Ethereum/ethereumjs-vm/node_modules/bn.js/lib/bn.js:6:21)
    at BN.toNumber (/Volumes/Ethereum/ethereumjs-vm/node_modules/bn.js/lib/bn.js:540:7)
    at LegacyTransaction._validateTxV (/Volumes/Ethereum/ethereumjs-vm/packages/tx/src/legacyTransaction.ts:267:30)
    at new LegacyTransaction (/Volumes/Ethereum/ethereumjs-vm/packages/tx/src/legacyTransaction.ts:109:12)
    at Function.fromValuesArray (/Volumes/Ethereum/ethereumjs-vm/packages/tx/src/legacyTransaction.ts:66:14)
    at Function.fromBlockBodyData (/Volumes/Ethereum/ethereumjs-vm/packages/tx/src/transactionFactory.ts:77:32)
    at Function.fromValuesArray (/Volumes/Ethereum/ethereumjs-vm/packages/block/src/block.ts:94:28)
    at /Volumes/Ethereum/ethereumjs-vm/packages/client/lib/sync/fetcher/blockfetcher.ts:59:20
    at Array.map (<anonymous>)
    at BlockFetcher.request (/Volumes/Ethereum/ethereumjs-vm/packages/client/lib/sync/fetcher/blockfetcher.ts:53:36)
    ```
    
Should be rather easy to fix though, we call `toNumber` on very large `v` values (this is caused by the high chain id).

**Comment by jochem-brouwer** [comment]
Tests are going to fail, because this PR now needs these changes to `ethereumjs-util`; https://github.com/ethereumjs/ethereumjs-util/pull/290

I can report that we are successfully syncing with the YoloV3 network. I am currently at block 33424, head is at 79783.

What a great milestone we have achieved here! 🎉 😄

**Comment by jochem-brouwer** [comment]
image

We synced to (almost) head of chain! We can use this opportunity to figure out what should happen at the head of the chain. It looks like the block fetcher stops once we are synced.

Note: to test this locally, pull the ethereumjs-util and checkout the `fix-ecrecover` branch. Now build it. Copy the `dist/signature*` files to `node_modules` of our monorepo (so the root) and overwrite the old signature files.

**Comment by jochem-brouwer** [comment]
Did a re-sync, I noticed that at some point I get this error

`Peer error: Error: Request timed out after 6000ms id=9e1096aa address=3.9.20.133:30303 transport=rlpx protocols=eth inbound=false`

(This is the bootnode). I wonder if we get banned from this peer, might be useful to take a look here.

**Comment by holgerd77** [comment]
Hi @jochem-brouwer, hope that was ok to merge in #1048, there was just so much other work heaping up! We'll continue to have a close eye on this of course and won't release anything prematurely, so don't worry! 😄 

What is the plan here? Do we want to cherry-pick the respective commits? What is the first relevant commit here?

**Comment by jochem-brouwer** [comment]
I will check this tomorrow but I think we can directly rebase this one and once we integrate the util package I think we can directly merge this 👍

**Comment by holgerd77** [comment]
Absolutely no chance for a rebase here - that's what I meant with warning here a bit 😋  - once you have done a rebase on a parent branch all the commits are different and git perceives (close to) all changes done as concurrent changes and you have to manually fuddle yourself through everything, which is close to not doable (just tested locally and it already completely breaks at the first commit of #1048 😄 ).

But nevermind, I think this really should be a good candidate for cherry-picking, will give this a try.

**Comment by holgerd77** [comment]
Ok, have cherry-picked the new commits from this branch and force-pushed. There is still an error occurring on the `yolov3` client run, generally this looks good though I would say.

Error is:

```shell
INFO [03-05|11:02:42] Started libp2p server.
INFO [03-05|11:02:45] Listener up transport=rlpx url=enode://c904debea44d4a797208a3a07994a80aa0e80293700a960a2c638827c1cf85f3e134d04d0f188ebbc5483430e8e06936a4163685ec5e922c42565f2567fae153@[::]:30303
INFO [03-05|11:02:45] Listener up transport=libp2p url=/ip4/127.0.0.1/tcp/50580/ws/p2p/QmPeAfmNKut6emwQqqqpa6xddg8oBPG8WDLDxZmfWj5go9
INFO [03-05|11:02:45] Started eth service.
INFO [03-05|11:02:54] Imported blocks count=50 number=1 hash=1053e007... hardfork=berlin peers=1
INFO [03-05|11:02:54] Executed blocks count=50 first=0 hash=f1f2876e... hardfork=berlin last=50 hash=f3d236ac... with txs=0
INFO [03-05|11:02:54] Executed blocks count=5 first=50 hash=f3d236ac... hardfork=berlin last=55 hash=4fe2c891... with txs=0
ERROR [03-05|11:02:54] Error: invalid transactions: errors at tx 0: Invalid Signature (block number=73 hash=c56fbbd792fac0da5c8d20112c317346b6e07586afe3b68ba866a184d6aa3ad0)     at BlockHeader._error (/ethereumjs-monorepo/packages/block/src/header.ts:762:15)
    at Block.validateData (/ethereumjs-monorepo/packages/block/src/block.ts:264:25)
    at Block.validate (/ethereumjs-monorepo/packages/block/src/block.ts:250:16)
    at /ethereumjs-monorepo/packages/blockchain/src/index.ts:880:9
    at Blockchain.runWithLock (/ethereumjs-monorepo/packages/blockchain/src/index.ts:407:21)
    at Blockchain._putBlockOrHeader (/ethereumjs-monorepo/packages/blockchain/src/index.ts:851:5)
    at Blockchain.putBlock (/ethereumjs-monorepo/packages/blockchain/src/index.ts:806:5)
    at Blockchain.putBlocks (/ethereumjs-monorepo/packages/blockchain/src/index.ts:792:7)
    at Chain.putBlocks (/ethereumjs-monorepo/packages/client/lib/blockchain/chain.ts:275:5)
    at BlockFetcher.store (/ethereumjs-monorepo/packages/client/lib/sync/fetcher/blockfetcher.ts:84:5)
```

Not sure if I have forgotten to cherry-pick a commit or this is something else. Have temporarily pushed the old branch to #1137 to ease debugging eventually.

**Comment by holgerd77** [comment]
Ok, this is now working again, will merge this in. 😄

**Review by holgerd77 — APPROVED** [review]
LGTM

**Pull request merged**